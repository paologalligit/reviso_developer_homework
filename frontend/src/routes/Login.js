import React, { Component } from 'react';
import { extendObservable } from 'mobx';
import { observer } from 'mobx-react';
import { Button, Container, Header, Form, Message } from 'semantic-ui-react';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { Link } from 'react-router-dom';

import styles from './styles/LoginStyle';
import isAuthenticated from '../utils/auth';

class Login extends Component {
  constructor(props) {
    super(props);

    extendObservable(this, {
      email: '',
      password: '',
      errors: {},
    });

    this.tryLogin();
  }

  tryLogin = () => {
    if (isAuthenticated()) {
      this.props.history.push('/');
    }
  };

  onChange = e => {
    const { name, value } = e.target;
    this[name] = value;
  };

  onSubmit = async () => {
    const { email, password } = this;

    const response = await this.props.mutate({
      variables: { email, password },
    });

    const { ok, token, refreshToken, errors } = response.data.login;

    if (ok) {
      localStorage.setItem('token', token);
      localStorage.setItem('refreshToken', refreshToken);

      this.props.history.push('/');
    } else {
      const err = {};
      errors.forEach(({ path, message }) => {
        err[`${path}Error`] = message;
      });

      this.errors = err;
    }
  };

  render() {
    const {
      email,
      password,
      errors: { emailError, passwordError },
    } = this;

    const errorList = [];

    if (emailError) {
      errorList.push(emailError);
    }

    if (passwordError) {
      errorList.push(passwordError);
    }

    console.log(errorList)

    return (
      <Container text textAlign={'center'} style={styles.login}>
        <Header as="h2">Login</Header>
        <Form>
          <Form.Field error={!!emailError}>
            <Form.Input
              icon="user"
              iconPosition="left"
              name="email"
              onChange={this.onChange}
              value={email}
              placeholder="Email"
              fluid
            />
          </Form.Field>
          <Form.Field error={!!passwordError}>
            <Form.Input
              icon="lock"
              iconPosition="left"
              name="password"
              onChange={this.onChange}
              value={password}
              type="password"
              placeholder="Password"
              fluid
            />
          </Form.Field>
          <Button onClick={this.onSubmit} primary size="huge">Submit</Button>
        </Form>
        {errorList.length ? (
          <Message
            error
            header="There was some errors with your submission"
            list={errorList}
            compact
          />
        ) : null}

        <Message>
          New to us? <Link to="/register">Sign Up</Link>
        </Message>
      </Container>
    );
  }
}

const loginMutation = gql`
  mutation($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      ok
      token
      refreshToken
      errors {
        path
        message
      }
    }
  }
`;

export default graphql(loginMutation)(observer(Login));
