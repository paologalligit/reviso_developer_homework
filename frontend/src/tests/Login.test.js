import React from 'react';
import toJson from 'enzyme-to-json';
import { mount, shallow, render } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

import Login from '../routes/Login';
import './setupTests';

describe('rendering', () => {
  const mocks = [];
  let wrapper;

  // initialization
  beforeEach(() => {
    wrapper = render(
      <MockedProvider mocks={mocks}>
        <Login />
      </MockedProvider>,
    );
  });

  it('matches snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders login page', () => {
    expect(wrapper).toHaveLength(1);
  });

  it('renders email and password text-input', () => {
    const textInputs = wrapper.find('input');

    expect(textInputs).toHaveLength(2);

    const email = textInputs[0];
    const passw = textInputs[1];

    expect(email.name).toBe('input');
    expect(email.attribs.name).toBe('email');
    expect(email.attribs.placeholder).toBe('Email');

    expect(passw.name).toBe('input');
    expect(passw.attribs.name).toBe('password');
    expect(passw.attribs.placeholder).toBe('Password');
  });


  it('renders submit button', () => {
    const button = wrapper.find('button');
    expect(button).toHaveLength(1);
  });

  it('initially should have flat state', () => {
    const login = shallow(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Login />
      </MockedProvider>,
    );

    const initialState = {
      email: '',
      password: '',
      errors: {},
    };

    login.setState({ initialState });

    expect(login.state().initialState).toEqual({
      email: '',
      password: '',
      errors: {},
    });
  });

  /*
  it('on click should call onSubmit and change state', () => {
    const button = shallow(wrapper.find('button'));
    const mockCallBack = jest.fn();

    button.simulate('click');
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
  */
});
