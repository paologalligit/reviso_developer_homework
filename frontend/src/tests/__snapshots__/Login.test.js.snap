// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login renders matches snapshot 1`] = `
<div
  class="ui text center aligned container"
  style="margin-top:50px"
>
  <h2
    class="ui header"
  >
    Login
  </h2>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <div
        class="field"
      >
        <div
          class="ui fluid left icon input"
        >
          <input
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
          <i
            aria-hidden="true"
            class="user icon"
          />
        </div>
      </div>
    </div>
    <div
      class="field"
    >
      <div
        class="field"
      >
        <div
          class="ui fluid left icon input"
        >
          <input
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <i
            aria-hidden="true"
            class="lock icon"
          />
        </div>
      </div>
    </div>
    <button
      class="ui huge primary button"
    >
      Submit
    </button>
  </form>
  <div
    class="ui message"
  >
    New to us? 
    <a
      href="/register"
    >
      Sign Up
    </a>
  </div>
</div>
`;
